File: src/2816.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "list.h"

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* doubleIt(struct ListNode* head) {
    struct ListNode* pRetVal = head;

    int doubleValue;
    struct ListNode* pPrevious = NULL;
    struct ListNode* pCurrent = head;
    while (pCurrent != NULL) {
        doubleValue = pCurrent->val * 2;
        if (doubleValue < 10) {
            pCurrent->val = doubleValue;
        } else if (pPrevious != NULL) {
            pCurrent->val = doubleValue % 10;
            pPrevious->val += 1;
        } else {
            head = (struct ListNode*)malloc(sizeof(struct ListNode));
            if (head == NULL) {
                perror("malloc");
                return pRetVal;
            }
            head->val = 1;
            head->next = pCurrent;
            pCurrent->val = doubleValue % 10;
        }
        pPrevious = pCurrent;
        pCurrent = pCurrent->next;
    }
    pRetVal = head;

    return pRetVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e4)
    struct testCaseType {
        int node[MAX_SIZE];
        int nodeSize;
    } testCase[] = {{{1, 8, 9}, 3}, {{9, 9, 9}, 3}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: head = [1,8,9]
     *  Output: [3,7,8]
     *
     *  Input: head = [9,9,9]
     *  Output: [1,9,9,8]
     */

    struct ListNode* pHead = NULL;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: ");
        for (j = 0; j < testCase[i].nodeSize; ++j) {
            pHead = addValueToEndListNode(pHead, testCase[i].node[j]);
        }
        printf("head = [");
        printList(pHead);
        printf("]\n");

        pHead = doubleIt(pHead);
        printf("Output: [");
        printList(pHead);
        printf("]\n");

        printf("\n");

        freeListNode(pHead);
        pHead = NULL;
    }

    return EXIT_SUCCESS;
}

File: src/2816.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop

from list import ListNode, ListNodeCommon


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        retVal = head

        previous = None
        current = head
        while current != None:
            doubleValue = current.val * 2
            if doubleValue < 10:
                current.val = doubleValue
            elif previous != None:
                current.val = doubleValue % 10
                previous.val += 1
            else:
                head = ListNode(1, current)
                current.val = doubleValue % 10
            previous = current
            current = current.next
        retVal = head

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for head in [[1, 8, 9], [9, 9, 9]]:
            # /* Example
            #  *  Input: head = [1,8,9]
            #  *  Output: [3,7,8]
            #  *
            #  *  Input: head = [9,9,9]
            #  *  Output: [1,9,9,8]
            #  */
            logging.info("Input: head = %s", head)

            head = ListNode.buildList(head)
            retVal = pSolution.doubleIt(head)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
